### PR TITLE
Add reverse IP lookups to pkg/fqdn.DNSCache

### PR DESCRIPTION
**Pull request by raybejjani**
It is sometimes useful to find all DNS names that have a specific IP. `pkg/fqdn.DNSCache` already does most of the bookkeeping for this and can maintain the additional reverse lookup information. This can then be used to ask cilium-agent for this via an API call (among other methods).

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6397)

### CODE REVIEW

**Comment by raybejjani** [comment]
test-me-please

File: pkg/fqdn/cache.go [hunk 1]
```diff
@@ -202,6 +252,34 @@ func (c *DNSCache) removeExpired(entries ipEntries, now time.Time) {
...
+			c.removeReverse(ip, entry)
 		}
 	}
 }
+
+// upsertReverse updates the reverse DNS cache for ip with entry, if it expires
+// later than the already-stored entry.
+// It is assumed that entry includes ip.
+func (c *DNSCache) upsertReverse(ip string, entry *cacheEntry) {
+	entries, exists := c.reverse[ip]
```
Comment (vadorovsky) [inline pkg/fqdn/cache.go:258]: `c.RLock`?

File: pkg/fqdn/cache.go [hunk 2]
```diff
@@ -202,6 +252,34 @@ func (c *DNSCache) removeExpired(entries ipEntries, now time.Time) {
...
+	}
+	entries[entry.Name] = entry
+}
+
+// removeReverse removes the reference between ip and the name stored in entry.
+// When no more refrences from ip to any name exist, the map entry is deleted
+// outright.
+// It is assumed that entry includes ip.
+func (c *DNSCache) removeReverse(ip string, entry *cacheEntry) {
+	entries, exists := c.reverse[ip]
```
Comment (vadorovsky) [inline pkg/fqdn/cache.go:272]: same here

File: pkg/fqdn/cache.go (hunk 1 above)
Comment (raybejjani) [reply pkg/fqdn/cache.go:258]: Both the `*Reverse` functions are called from places that already hold the lock, no? Did I miss some path?

File: pkg/fqdn/cache.go [hunk 3]
```diff
@@ -181,6 +193,43 @@ func (c *DNSCache) lookupByRegexpByTime(now time.Time, re *regexp.Regexp) (match
...
+
+// lookupIPByTime takes a timestamp for expiration comparisions, and is
+// only intended for testing.
+func (c *DNSCache) lookupIPByTime(now time.Time, ip net.IP) (names []string) {
+	// find ip entry in reverse
+	// return name & entries that are not expired
+
+	ipKey := ip.String()
+
+	c.RLock()
```
Comment (tgraf) [inline pkg/fqdn/cache.go]: Double lock via `Lookup()`

File: pkg/fqdn/cache.go (hunk 1 above)
Comment (tgraf) [reply pkg/fqdn/cache.go:258]: The caller has the lock requirement in the description but not this function itself which makes this not obvious. Just add a comment to the function description

**Review by tgraf — CHANGES_REQUESTED** [review]

File: pkg/fqdn/cache.go (hunk 1 above)
Comment (vadorovsky) [reply pkg/fqdn/cache.go:258]: Sorry, you're right.

**Comment by raybejjani** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]

**Pull request merged**